File: split/makescripts.pas
procedure makescripts(q: halfword; delta: scaled);
var
  p, x, y, z: halfword;
  shiftup, shiftdown, clr: scaled;
  t: smallnumber;
begin
  p := mem[q + 1].int;
  if (p >= himemmin) then
  begin
    shiftup := 0;
    shiftdown := 0;
  end
  else
  begin
    z := hpack(p, 0, 1);
    if curstyle < 4 then
      t := 16
    else
      t := 32;
    shiftup := mem[z + 3].int - fontinfo[18 + parambase[eqtb[3937 + t].hh.rh]].int;
    shiftdown := mem[z + 2].int + fontinfo[19 + parambase[eqtb[3937 + t].hh.rh]].int;
    freenode(z, 7);
  end;
  if mem[q + 2].hh.rh = 0 then{757:}

  begin
    x := cleanbox(q + 3, 2 * (curstyle div 4) + 5);
    mem[x + 1].int := mem[x + 1].int + eqtb[5842].int;
    if shiftdown < fontinfo[16 + parambase[eqtb[3937 + cursize].hh.rh]].int then
      shiftdown := fontinfo[16 + parambase[eqtb[3937 + cursize].hh.rh]].int;
    clr := mem[x + 3].int - (abs(fontinfo[5 + parambase[eqtb[3937 + cursize].hh.rh]].int * 4) div 5);
    if shiftdown < clr then
      shiftdown := clr;
    mem[x + 4].int := shiftdown;
  end{:757}

  else
  begin{758:}

    begin
      x := cleanbox(q + 2, 2 * (curstyle div 4) + 4 + (curstyle mod 2));
      mem[x + 1].int := mem[x + 1].int + eqtb[5842].int;
      if odd(curstyle) then
        clr := fontinfo[15 + parambase[eqtb[3937 + cursize].hh.rh]].int
      else if curstyle < 2 then
        clr := fontinfo[13 + parambase[eqtb[3937 + cursize].hh.rh]].int
      else
        clr := fontinfo[14 + parambase[eqtb[3937 + cursize].hh.rh]].int;
      if shiftup < clr then
        shiftup := clr;
      clr := mem[x + 2].int + (abs(fontinfo[5 + parambase[eqtb[3937 + cursize].hh.rh]].int) div 4);
      if shiftup < clr then
        shiftup := clr;
    end{:758}
;
    if mem[q + 3].hh.rh = 0 then
      mem[x + 4].int := -shiftup
    else{759:}

    begin
      y :=
        cleanbox(q + 3, 2 * (curstyle div 4) + 5);
      mem[y + 1].int := mem[y + 1].int + eqtb[5842].int;
      if shiftdown < fontinfo[17 + parambase[eqtb[3937 + cursize].hh.rh]].int then
        shiftdown := fontinfo[17 + parambase[eqtb[3937 + cursize].hh.rh]].int;
      clr := 4 * fontinfo[8 + parambase[eqtb[3938 + cursize].hh.rh]].int - ((shiftup - mem[x + 2].int) - (mem[y + 3].int - shiftdown));
      if clr > 0 then
      begin
        shiftdown := shiftdown + clr;
        clr := (abs(fontinfo[5 + parambase[eqtb[3937 + cursize].hh.rh]].int * 4) div 5) - (shiftup - mem[x + 2].int);
        if clr > 0 then
        begin
          shiftup := shiftup + clr;
          shiftdown := shiftdown - clr;
        end;
      end;
      mem[x + 4].int := delta;
      p := newkern((shiftup - mem[x + 2].int) - (mem[y + 3].int - shiftdown));
      mem[x].hh.rh := p;
      mem[p].hh.rh := y;
      x := vpackage(x, 0, 1, 1073741823);
      mem[x + 4].int := shiftdown;
    end{:759}
;
  end;
  if mem[q + 1].int = 0 then
    mem[q + 1].int := x
  else
  begin
    p := mem[q + 1].int;
    while mem[p].hh.rh <> 0 do
      p := mem[p].hh.rh;
    mem[p].hh.rh := x;
  end;
end;

File: cpp/makescripts.cpp
#include "makescripts.h"
#include "hpack.h"
#include "cleanbox.h"
#include "freenode.h"
#include "newkern.h"
#include "vpackage.h"

static int script_space(void) { return dimen_par(script_space_code); }

//! superscript shift-up in uncramped display style
static int sup1(smallnumber c) { return mathsy(13, c); }

//! superscript shift-up in uncramped non-display
static int sup2(smallnumber c) { return mathsy(14, c); }

//! superscript shift-up in cramped styles
static int sup3(smallnumber c) { return mathsy(15, c); }

//! subscript shift-down if superscript is absent
static int sub1(smallnumber c) { return mathsy(16, c); }

//! subscript shift-down if superscript is present
static int sub2(smallnumber c) { return mathsy(17, c); }

//! superscript baseline below top of large box
static int sup_drop(smallnumber c) { return mathsy(18, c); }

//! subscript baseline below bottom of large box
static int sub_drop(smallnumber c) { return mathsy(19, c); }

void makescripts(halfword q, scaled delta)
{
	halfword p = new_hlist(q);
	scaled shiftup, shiftdown;
	if (is_char_node(p))
	{
		shiftup = 0;
		shiftdown = 0;
	}
	else
	{
		auto z = hpack(p, 0, additional);
		smallnumber t = curstyle < script_style ? script_size : script_script_size;
		shiftup = height(z)-sup_drop(t);
		shiftdown = depth(z)+sub_drop(t);
		freenode(z, box_node_size);
	}
	scaled clr;
	halfword x;
	if (math_type(supscr(q)) == 0)
	{
		x = cleanbox(subscr(q), sub_style(curstyle));
		width(x) += script_space();
		if (shiftdown < sub1(cursize))
			shiftdown = sub1(cursize);
		clr = height(x)-(math_x_height(cursize)*4)/5;
		if (shiftdown < clr)
			shiftdown = clr;
		shift_amount(x) = shiftdown;
	}
	else
	{
		x = cleanbox(supscr(q), sup_style(curstyle));
		width(x) += script_space();
		if (curstyle%2)
			clr = sup3(cursize);
		else 
			if (curstyle < 2)
				clr = sup1(cursize);
			else
				clr = sup2(cursize);
		if (shiftup < clr)
			shiftup = clr;
		clr = depth(x)+abs(math_x_height(cursize))/4;
		if (shiftup < clr)
			shiftup = clr;
		if (math_type(subscr(q)) == 0)
			shift_amount(x) = -shiftup;
		else
		{
			auto y = cleanbox(subscr(q), sub_style(curstyle));
			width(y) += script_space();
			if (shiftdown < sub2(cursize))
				shiftdown = sub2(cursize);
			clr = 4*default_rule_thickness()-(shiftup-depth(x)-(height(y)-shiftdown));
			if (clr > 0)
			{
				shiftdown += clr;
				clr = (abs(math_x_height(cursize)*4)/5)-(shiftup-depth(x));
				if (clr > 0)
				{
					shiftup += clr;
					shiftdown -= clr;
				}
			}
			shift_amount(x) = delta;
			p = newkern(shiftup-depth(x)-(height(y)-shiftdown));
			link(x) = p;
			link(p) = y;
			x = vpack(x, 0, additional);
			shift_amount(x) = shiftdown;
		}
	}
	if (new_hlist(q) == 0)
		new_hlist(q) = x;
	else
	{
		p = new_hlist(q);
		while (link(p))
			p = link(p);
		link(p) = x;
	}
}
